### ISSUE TITLE
Add an example for Stage and Scene with a custom context

### ISSUE THREAD

**Issue by negebauer** [telegraf/telegraf#1139]
**Is your feature request related to a problem? Please describe.**

I'm new to typescript and I've been trying to write a bot that has uses session with a custom session object and stages.
I haven't found any examples of Stages using a custom context type.

**Describe the solution you'd like**

A complete example with a bot with a custom type and stages using that custom type.

**Additional context**

This is the code I have

```ts
import { Telegraf, Context, Stage, BaseScene } from 'telegraf'
import LocalSession from 'telegraf-session-local'

interface Session {
  bar: string
}

interface BotContext extends Context {
  session: Session
}

const bot = new Telegraf<BotContext>('foo')
const session = new LocalSession<BotContext>()
bot.use(session.middleware())

const scene = new BaseScene('scene')
scene.command('back', ctx => ctx.scene.leave())

const stage = new Stage([scene], { ttl: 10 })
bot.use(stage.middleware())

bot.command('foo', ctx => ctx.scene.enter('scene'))

bot.launch()
```

`bot.use(stage.middleware())` shows the following

```
Argument of type 'MiddlewareFn<SceneContextMessageUpdate>' is not assignable to parameter of type 'Middleware<BotContext>'.
  Type 'MiddlewareFn<SceneContextMessageUpdate>' is not assignable to type 'MiddlewareFn<BotContext>'.
    Property 'scene' is missing in type 'BotContext' but required in type 'SceneContextMessageUpdate'.
```

and `bot.command('foo', ctx => ctx.scene.enter('scene'))` shows

```
Property 'scene' does not exist on type 'BotContext'.
```

I though that using `BaseScene<BotContext>` and `Stage<BotContext>` would make it work, but I get the following instead

For `BaseScene<BotContext>`

```
Type 'BotContext' does not satisfy the constraint 'SceneContextMessageUpdate'.
  Property 'scene' is missing in type 'BotContext' but required in type 'SceneContextMessageUpdate'.
```

And for `Stage<BotContext>`

```
Type 'BotContext' does not satisfy the constraint 'SceneContextMessageUpdate'.
```

***

I did get to code that compiles by doing the following

```ts
import { Telegraf, Context, Stage, BaseScene } from 'telegraf'
import LocalSession from 'telegraf-session-local'
import { SceneContextMessageUpdate } from '../node_modules/telegraf/typings/stage.d'

interface Session {
  foo: string
}

type BaseBotContext = Context & SceneContextMessageUpdate

interface BotContext extends BaseBotContext {
  session: Session
}

const bot = new Telegraf<BotContext>('foo')
const session = new LocalSession<BotContext>()
bot.use(session.middleware())

const scene = new BaseScene<BotContext>('scene')
scene.command('back', ctx => ctx.scene.leave())

const stage = new Stage<BotContext>([scene], { ttl: 10 })
bot.use(stage.middleware())

bot.command('foo', ctx => ctx.scene.enter('scene'))

bot.launch()
```

Is the `type BaseBotContext = Context & SceneContextMessageUpdate` the way to do this?

**Comment by Morb0** [comment]
Thanks for sharing the solution.
Is it possible to somehow use `Stage` factories like `enter`, `leave`?
Now I got error: `Argument of type 'Middleware<SceneContextMessageUpdate>' is not assignable to parameter of type 'Middleware<Context>'`

**Comment by negebauer** [comment]
I can use `enter` and `leave` like this

```ts
import { Telegraf, Context, Stage, BaseScene } from 'telegraf'
import LocalSession from 'telegraf-session-local'
import { SceneContextMessageUpdate } from '../node_modules/telegraf/typings/stage.d'

const { enter, leave } = Stage

interface Session {
  foo: string
}

type BaseBotContext = Context & SceneContextMessageUpdate

interface BotContext extends BaseBotContext {
  session: Session
}

const bot = new Telegraf<BotContext>('foo')
const session = new LocalSession<BotContext>()
bot.use(session.middleware())

const scene = new BaseScene<BotContext>('scene')
scene.command('back', ctx => ctx.scene.leave())
scene.hears('leave', () => {
  leave()
})

const stage = new Stage<BotContext>([scene], { ttl: 10 })
bot.use(stage.middleware())

bot.hears('enter', () => {
  enter('scene')
})

bot.command('foo', ctx => ctx.scene.enter('scene'))

bot.launch()
```

This code should make `BotContext` satisfy `SceneContextMessageUpdate`

```ts
import { Context } from 'telegraf'
import { SceneContextMessageUpdate } from '../node_modules/telegraf/typings/stage.d'

type BaseBotContext = Context & SceneContextMessageUpdate

interface BotContext extends BaseBotContext {
  // your custom context specification
}
```ts

**Comment by Morb0** [comment]
The example you provided works fine.
But if use a scene with this `BotContext`, then Typescript gives an error:
`Type 'Context' does not satisfy the constraint 'SceneContextMessageUpdate'`

Example:
```typescript
let scene: BaseScene<BotContext (<- here error)>;
scene = new BaseScene('id');
```

**Comment by negebauer** [comment]
Dunno why this, which makes `scene` be of type `BaseScene<BotContext>`, works

```ts
const scene = new BaseScene<BotContext>('id')
```

But the example you provide doesn't.

I'm new to typescript, I don't what's the correct way to do this 😕

**Comment by SoundAsleep192** [comment]
`const bot = new Telegraf<SceneContextMessageUpdate>('foo')` did the trick for me

**Comment by KnorpelSenf** [comment]
Fixed in #1194
